src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/SnsService.kt

```diff
@@ -0,0 +1,98 @@
+import software.amazon.awssdk.services.sns.model.PublishRequest
+import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.service.AuditType
+import uk.gov.justice.hmpps.sqs.HmppsQueueService
+import java.time.Instant
+import java.time.LocalDateTime
+import java.time.ZoneId
+import java.time.format.DateTimeFormatter
+
+@Service
+class SnsService(hmppsQueueService: HmppsQueueService, private val objectMapper: ObjectMapper) {
+  companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  private val domaineventsTopic by lazy {
+    hmppsQueueService.findByTopicId("domainevents")
+      ?: throw RuntimeException("Topic with name domainevents doesn't exist")
+  }
+  private val domaineventsTopicClient by lazy { domaineventsTopic.snsClient }
+
+  @WithSpan(value = "hmpps-domain-events-topic", kind = SpanKind.PRODUCER)
+  fun publishDomainEvent(
+    eventType: NonAssociationDomainEventType,
+    description: String,
+    occurredAt: LocalDateTime,
+    additionalInformation: AdditionalInformation? = null,
+  ) {
+    publishToDomainEventsTopic(
+      HMPPSDomainEvent(
+        eventType.value,
+        additionalInformation,
+        occurredAt.atZone(ZoneId.systemDefault()).toInstant(),
+        description,
+      ),
+    )
+  }
+
+  private fun publishToDomainEventsTopic(payload: HMPPSDomainEvent) {
+    log.debug("Event {} for id {}", payload.eventType, payload.additionalInformation)
+    domaineventsTopicClient.publish(
+      PublishRequest.builder()
+        .topicArn(domaineventsTopic.arn)
+        .message(objectMapper.writeValueAsString(payload))
+        .messageAttributes(
+          mapOf(
+            "eventType" to MessageAttributeValue.builder().dataType("String").stringValue(payload.eventType).build(),
+          ),
+        )
+        .build()
+        .also { log.info("Published event $payload to outbound topic") },
+    )
+  }
+}
+
+data class AdditionalInformation(
+  val nonAssociationId: String? = null,
+  val prisonerNumber: String? = null,
+  val reason: String? = null,
+  val prisonId: String? = null,
+)
+
+data class HMPPSDomainEvent(
+  val eventType: String? = null,
+  val additionalInformation: AdditionalInformation?,
+  val version: String,
+  val occurredAt: String,
+  val description: String,
+) {
+  constructor(
+    eventType: String,
+    additionalInformation: AdditionalInformation?,
+    occurredAt: Instant,
+    description: String,
+  ) : this(
+    eventType,
+    additionalInformation,
+    "1.0",
+    occurredAt.toOffsetDateFormat(),
+    description,
+  )
+}
+
+enum class NonAssociationDomainEventType(val value: String, val description: String, val auditType: AuditType) {
+  NON_ASSOCIATION_CREATED("non-association.created", "A non-association has been created: ", AuditType.NON_ASSOCIATION_CREATED),
+  NON_ASSOCIATION_UPSERT("non-association.amended", "A non-association has been amended: ", AuditType.NON_ASSOCIATION_UPDATED),
```

Comment:
namespace should be plural, e.g. `non-associations.*`

I'd also call the event `NON_ASSOCIATION_UPDATED("non-associations.updated", ...`

Sorry, don't mean to be too picky, we could always rename later on but it's a bit tricker with events stuff.

Reply:
I've updated to plural but I think the convention is to used created and amended now

Reply:
Really? That's odd, alright I guess 😅